pkg/client/chaosclient/chaosclient.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package chaosclient makes it easy to simulate network latency, misbehaving
+// servers, and random errors from servers. It is intended to stress test components
+// under failure conditions and expose weaknesses in the error handling logic
+// of the codebase.
+package chaosclient
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"reflect"
+	"runtime"
+)
+
+// chaosrt provides the ability to perform simulations of HTTP client failures
+// under the Golang http.Transport interface.
+type chaosrt struct {
+	rt     http.RoundTripper
+	notify ChaosNotifier
+	c      []Chaos
+}
+
+// Chaos intercepts requests to a remote HTTP endpoint and can inject arbitrary
+// failures.
+type Chaos interface {
+	// Intercept should return true if the normal flow should be skipped, and the
+	// return response and error used instead. Modifications to the request will
+	// be ignored, but may be used to make decisions about types of failures.
+	Intercept(req *http.Request) (bool, *http.Response, error)
```

Comment:
I remember now - we also want to emulate errors, so the return code would check (response != nil || err != nil) which is sufficiently un-Golike that I felt the bool was more accurate.

----- Original Message -----
> In general in Go I prefer not to overload the meaning of nil, but in this
> case I don't think it would be an issue.
> 
> ----- Original Message -----
> > > +
> > > +// chaosrt provides the ability to perform simulations of HTTP client
> > > failures
> > > +// under the Golang http.Transport interface.
> > > +type chaosrt struct {
> > > +	rt     http.RoundTripper
> > > +	notify ChaosNotifier
> > > +	c      []Chaos
> > > +}
> > > +
> > > +// Chaos intercepts requests to a remote HTTP endpoint and can inject
> > > arbitrary
> > > +// failures.
> > > +type Chaos interface {
> > > +	// Intercept should return true if the normal flow should be skipped,
> > > and
> > > the
> > > +	// return response and error used instead. Modifications to the request
> > > will
> > > +	// be ignored, but may be used to make decisions about types of
> > > failures.
> > > +	Intercept(req *http.Request) (bool, *http.Response, error)
> > 
> > Is the "bool" return value needed? You could have the presence of the
> > Response object be the signal to override the response.
> > 
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/GoogleCloudPlatform/kubernetes/pull/6729/files#r28261402
>